test/unit-test/mqtt_agent_utest.c

```diff
@@ -74,21 +110,449 @@ int suiteTearDown( int numFailures )
+                                 mockPublishCallback,
+                                 incomingPacketContext );
+    TEST_ASSERT_EQUAL( MQTTSuccess, mqttStatus );
+    /* Set packet ID nonzero to indicate initialization. */
+    pAgentContext->mqttContext.nextPacketId = 1U;
+}
+
+/* ========================================================================== */
+
+/**
+ * @brief Test that MQTTAgent_Init is able to update the context object correctly.
+ */
+
+void test_MQTTAgent_Init_Happy_Path( void )
+{
 }
 
 /**
  * @brief Test that any NULL parameter causes MQTTAgent_Init to return MQTTBadParameter.
  */
 void test_MQTT_Init_Invalid_Params( void )
 {
-    
+}
+
+/* ========================================================================== */
+
+/**
+ * @brief Test that the parameter validation for API functions that create
+ * commands (publish, subscribe, connect, etc.) works as intended.
+ */
+void test_MQTTAgent_API_Create_Command_Invalid_Params( void )
+{
+    MQTTAgentContext_t agentContext = { 0 };
+    MQTTStatus_t mqttStatus;
+    CommandInfo_t commandInfo = { 0 };
+
+    setupAgentContext( &agentContext );
+
+    /* Use Terminate since it's one that doesn't have an additional pointer param. */
+    mqttStatus = MQTTAgent_Terminate( &agentContext, NULL );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, mqttStatus );
+
+    mqttStatus = MQTTAgent_Terminate( NULL, &commandInfo );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, mqttStatus );
+
+    /* Various NULL parameters for the agent interface. */
+    agentContext.agentInterface.send = NULL;
+    mqttStatus = MQTTAgent_Terminate( &agentContext, &commandInfo );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, mqttStatus );
+
+    agentContext.agentInterface.send = mockSend;
+    agentContext.agentInterface.recv = NULL;
+    mqttStatus = MQTTAgent_Terminate( &agentContext, &commandInfo );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, mqttStatus );
+
+    agentContext.agentInterface.recv = mockReceive;
+    agentContext.agentInterface.getCommand = NULL;
+    mqttStatus = MQTTAgent_Terminate( &agentContext, &commandInfo );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, mqttStatus );
+
+    agentContext.agentInterface.getCommand = mockGetCommand;
+    agentContext.agentInterface.releaseCommand = NULL;
+    mqttStatus = MQTTAgent_Terminate( &agentContext, &commandInfo );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, mqttStatus );
+
+    agentContext.agentInterface.releaseCommand = mockReleaseCommand;
+    agentContext.agentInterface.pMsgCtx = NULL;
+    mqttStatus = MQTTAgent_Terminate( &agentContext, &commandInfo );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, mqttStatus );
+
+    agentContext.agentInterface.pMsgCtx = &globalMessageContext;
+    /* Invalid packet ID to indicate uninitialized context. */
+    agentContext.mqttContext.nextPacketId = MQTT_PACKET_ID_INVALID;
+    mqttStatus = MQTTAgent_Terminate( &agentContext, &commandInfo );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, mqttStatus );
```

Comment:
Why is this testing limited to the MQTTAgent_Terminate API for create command testing? Is it testing all the paths for bad parameters for create command?

Reply:
Every single thread-safe API function is a wrapper around the same static function to create commands and add to the queue. I thought it redundant to test these cases for all 8 of them, so I just tested this one exhaustively and the others for coverage. Terminate just happened to be the one I picked

Reply:
This seems like we are using the knowledge of the implementation to decide which API to test to achieve coverage numbers . Tests should be written against APIs - what if I update one `MQTTAgent_` API and stop calling the common function which checks for error. Would it introduced an error without failing a test.